olp-cpp-sdk-core/src/network2/NetworkCurl.cpp

```diff
@@ -0,0 +1,410 @@
+  NetworkRequest request;
+  std::shared_ptr<std::ostream> payload;
+  Network::Callback callback;
+  Network::HeaderCallback header_callback;
+  Network::DataCallback data_callback;
+
+  // fields used by internal worker thread
+  CURL* curl_handle;
+  curl_slist* header_list;
+  size_t read_cursor;
+  size_t write_cursor;
+};
+
+class CurlNetwork final : public Network {
+ public:
+  CurlNetwork(int connection_pool_size);
+  virtual ~CurlNetwork() override;
+
+  RequestId Send(NetworkRequest request, std::shared_ptr<std::ostream> payload,
+                 Callback callback, HeaderCallback header_callback = nullptr,
+                 DataCallback data_callback = nullptr) override;
+  void Cancel(RequestId id) override;
+
+ private:
+  void NotifyWorkerThread();
+
+  // methods used by internal worker thread
+  void Run();  /// worker thread starts here
+  void FlushWorkerNotifications();
+  void StartNewTransfers();
+  void HandleCancelledTransfers();
+  void ProcessFinishedTransfers();
+
+  std::thread worker_;
+
+  CURLM* curl_multi_handle_;
+  std::vector<CURL*> free_handles_;
+  curl_waitfd curl_custom_wait_;
+
+  std::map<CURL*, std::unique_ptr<RequestContext>> running_requests_;
+  std::map<RequestId, CURL*>
+      running_requests_index_;  /// helps to find request context by request id
+
+  /// these members are accessed by both clients and worker thread
+  std::mutex mutex_;
+  RequestId next_request_id_ = RequestIdConstants::RequestIdMin;
+  int worker_notification_pipe_[2];
+  std::atomic<bool> running_;
+  std::map<RequestId, std::unique_ptr<RequestContext>> request_queue_;
+  std::vector<RequestId> cancel_list_;
+};
+
+CurlNetwork::CurlNetwork(int connection_pool_size) {
+  curl_multi_handle_ = curl_multi_init();
+  for (int i = 0; i < connection_pool_size; i++) {
+    free_handles_.emplace_back(curl_easy_init());
+  }
+
+  pipe(worker_notification_pipe_);
+  fcntl(worker_notification_pipe_[0], F_SETFL, O_NONBLOCK);
+  fcntl(worker_notification_pipe_[1], F_SETFL, O_NONBLOCK);
+  curl_custom_wait_.fd = worker_notification_pipe_[0];
+  curl_custom_wait_.events = CURL_WAIT_POLLIN;
+  curl_custom_wait_.revents = 0;
+
+  running_.store(true);
+  worker_ = std::thread(&CurlNetwork::Run, this);
+}
+
+CurlNetwork::~CurlNetwork() {
+  running_.store(false);
+  NotifyWorkerThread();
+  worker_.join();
+
+  close(worker_notification_pipe_[0]);
+  close(worker_notification_pipe_[1]);
+
+  for (auto easy_handle : free_handles_) {
+    curl_easy_cleanup(easy_handle);
+  }
+  curl_multi_cleanup(curl_multi_handle_);
+}
+
+void CurlNetwork::FlushWorkerNotifications() {
+  if (!(curl_custom_wait_.revents & CURL_WAIT_POLLIN)) {
+    return;
+  }
+  char unused;
+  while (read(worker_notification_pipe_[0], &unused, sizeof(unused)) > 0)
+    ;
+  return;
+}
+
+size_t curl_write_callback(char* ptr, size_t size, size_t nmemb,
+                           void* userdata) {
+  RequestContext& context = *reinterpret_cast<RequestContext*>(userdata);
```

Comment:
assert userdata for non-zero?

Reply:
If only then nullptr checks and log/return.
Aborting/assert on edge devices is a serious sin and almost always disabled.